### DISCUSSION TITLE
Lua Support

### DISCUSSION THREAD

**Discussion by magnusviri** [endless-sky/endless-sky#10279] [Ideas]
I forked this repo and added [branches](https://github.com/magnusviri/endless-sky/branches) to my fork that include [Koranir's](https://github.com/Koranir) [Lua pull request](https://github.com/endless-sky/endless-sky/pull/9504) but I updated it for ES 0.10.6, 0.10.7, and 0.10.8. Starting with my [lua-0.10.8](https://github.com/magnusviri/endless-sky/tree/lua-0.10.8) branch I've started adding my own contributions, taking it farther than Koranir did.

I am new to this community and see [several](https://github.com/endless-sky/endless-sky/pull/9504#issuecomment-1826149852) [comments](https://github.com/endless-sky/endless-sky/issues/1841#issuecomment-262980070) saying that Lua support isn't desired but then there's also some people who appear to want it. So I'm kind of walking on my tip toes here because I don't know how what I'm doing will be received.

My goal in working on this is so I can hook up an LLM like Llama 3 to the game. Specifically, I want to generate "missions" and dialog on-the-fly using an LLM. I have been playing my own scifi LLM game with a universe that is remarkably similar to this game's universe. I decided I needed a game engine that supplied the LLM with data and rules and I started to write my own but then I found this game. I've been playing ES for a few weeks or so and I really love it. But I want to get back to the LLM stuff I was doing also. When I first found this game I decided I would get more bang for the buck if I added Lua support to ES and use Lua to hook up to my LLM than for me to try to write my own game engine. There's no way I could come close to the experience of ES's gameplay. So that is what I've been doing. (I'm not good enough with c++ to hook up my LLM that way.)

If you haven't looked at Koranir's pull request, basically it all starts with the existing plugin system. He added the ["script" keyword](https://github.com/magnusviri/endless-sky/blob/lua-0.10.8/data/luatest.txt) to load a [Lua script](https://github.com/magnusviri/endless-sky/blob/lua-0.10.8/scripts/luatest.lua). Then he added a [hook system](https://github.com/magnusviri/endless-sky/blob/13f8d4a9988137e208088f9b15c7c0bf5931a624/source/LuaPlugin.cpp) so the c++ code can look in the Lua scripts for Lua functions and run them if they exist. He only added two hooks, "[es_init](https://github.com/magnusviri/endless-sky/blob/13f8d4a9988137e208088f9b15c7c0bf5931a624/source/PlayerInfo.cpp#L137)" and "[es_daily](https://github.com/magnusviri/endless-sky/blob/13f8d4a9988137e208088f9b15c7c0bf5931a624/source/PlayerInfo.cpp#L794)". Then he added the ability for the Lua code to [write messages to the game window](https://github.com/magnusviri/endless-sky/blob/13f8d4a9988137e208088f9b15c7c0bf5931a624/source/LuaImpl.cpp#L28). I added a third hook, "es_add_crew". I did this because in my LLM game, interacting with the crew is pretty much most of the game. I saw a Firefly-class transport ship in ES, so I know someone in the ES community knows how interesting a crew can be. I added my hook to [Ship::AddCrew](https://github.com/magnusviri/endless-sky/blob/13f8d4a9988137e208088f9b15c7c0bf5931a624/source/Ship.cpp#L2935). I check to see if the ship IsYours() first since enemy ships can have their crews negatively "added to" when there's combat.

Basically, I'm hoping the dev community can tell me what are good places to add hooks and where to avoid them because the existing plugin system already does what is needed. Here's the types of things I'm hoping to do. I want to create dialog based missions that trigger when the player visits certain planets or areas on the planet, like landing, the spaceport, the bank, the outfitter, etc. I'd also like to add the ability to just go explore a planet via dialog (probably when visiting the spaceport). I know there's a lot of code already that does this for the missions, like, I've been chasing after dragons and went looking for treasure in the ocean. So it's there, I just need to figure out how to hook it up to Lua to generate on-the-fly. I need to be able to save and load the Lua plugin state. I figure there's already code that will save and load plugin parameters but I haven't figured out how that works.

Another thing I thought would be cool is if every time I leave a planet I can have one of my crew members just fill the cargo hold with the most profitable cargo based on my destination and then sell it when I reach my destination. Obviously, this crew member will get paid more than normal, maybe even taking a cut of the profits. Otherwise this could remove one of the core game mechanics too early in the game (at some point in the game I quit buying cargo because I could earn so many credits so quickly by boarding ships and other high profit missions that cargo wasn't worth my time anymore--but it would be worth paying a crew member half the profits if my crew member did the work for me). So I would need a hook that runs right before take off and right when landing.

You know, this is where having an LLM would make this so much fun. It would be great if a Lua plugin tracked how honest and loyal my crew was. So I could have a corrupt crew member that took more than his share of the profits. Or maybe the crew member started trafficking illegal goods without my knowledge. If I hired crew from pirate planets maybe they would murder the crew members that were hired on the "soft" rich worlds. Or one of my escort ships might start firing on another escort. And of course, there could be mutinies. This is where an LLM would make things really interesting. Programming something like this the old way, you know, structured code, would not be feasible. But a mutiny is driven by human emotions, and ironically, LLMs are really good at navigating human emotions.

Anyway, that brings me to the next thing, which is the Lua scripts need to be able to get game state. I've exposed the player object to Lua using [LuaBridge3](https://github.com/kunitoki/LuaBridge3). That happens in [main.cpp](https://github.com/magnusviri/endless-sky/blob/13f8d4a9988137e208088f9b15c7c0bf5931a624/source/main.cpp#L288) and [Lua.cpp](https://github.com/magnusviri/endless-sky/blob/13f8d4a9988137e208088f9b15c7c0bf5931a624/source/Lua.cpp#L237). After playing with LuaBridge3 a while and finally being able to change the players name, view the system and planet the player is on, and add credits, I decided to take a scorched earth approach and I asked Claude to write a perl script that reads all of the ES headers and write LuaBridge code to expose as much as possible to Lua. You can view that all in [Lua.cpp](https://github.com/magnusviri/endless-sky/blob/13f8d4a9988137e208088f9b15c7c0bf5931a624/source/Lua.cpp#L251). It's messy, not tested, and can only be accessed if the player object can access it, so yeah it needs a ton of cleanup, but it compiles (after commenting out a lot). ES has a lot of overloading and LuaBridge3 doesn't work with overloading. But Claude also was able to get around it so I can [access the player's Account](https://github.com/magnusviri/endless-sky/blob/13f8d4a9988137e208088f9b15c7c0bf5931a624/source/Lua.cpp#L1520).

I'm using LuaBridge3 because it makes this so much easier and allows me to determine if any of this is even a good idea. If any of this stuff were ever going to be merged into the main repo, I would assume that it would all be switched to raw Lua, similar to the [printMsg function](https://github.com/magnusviri/endless-sky/blob/13f8d4a9988137e208088f9b15c7c0bf5931a624/source/LuaImpl.cpp#L28). I wouldn't dare do this by hand. There's so much to do, it only makes sense to have a script that reads the source code and auto-generates the Lua code. I mean, that's basically what LuaBridge does anyway, it just does it at runtime instead of compile time.

I'm not very good with c++. I can read it and imitate and repurpose code, if I can find the right examples. I'm not a fulltime developer and have only tinkered with game dev. I'm pretty familiar with [Love2d](https://www.love2d.org/), which essentially exposes SDL to Lua. I can figure out a lot of stuff with diffs and grep. And I've been able to use ChatGPT/Claude to do far more than my skill level if I were on my own. I have started reading ES's plugin system and how to generate content. And I've gone through a lot of the source code and I have a general idea how the game works. But there's still too much I don't know.

So finally, if you have ideas of where to add hooks and what data needs to be exposed to Lua, please tell me here. And if you have answers to my questions, like how should I go about saving/loading the Lua data to the save file, and how I could generate missions on-the-fly with Lua, please let me know. And if there's things to do or things to avoid to increase the chance of this being accepted to the main repo, I will try to do those things.

**Comment by Saugia** [comment]
Personally I don't know much about Lua and coding myself, but https://github.com/endless-sky/endless-sky/pull/9504 would be worth reading a little bit on to take a look at what previous comments and thoughts on Lua support have been in the past; primarily, this comment from the original developer of the game https://github.com/endless-sky/endless-sky/issues/1841#issuecomment-262980070 and the issue it's found in. I can't vouch for if the thought remains the same, so I would wait until other coders and developers take a look at what you have to show us here, since it's quite extensively discussed.

Edit: I do see you've acknowledged that Koranir did attempt to add this in the past, my bad.

**Comment by lumbar527** [comment]
I think a way to implement scripting inside ES would be very useful, allowing for much more customization.

I think it would also be useful if missions/events could trigger scripts.

I've used Lua some through [Pico-8](https://www.lexaloffle.com/pico-8.php) and it's quite a simple language. I've been a programmer for a while so I'm not sure how non-programmers would take to this but the added difficulty of it would be the programming logic.

**Reply by magnusviri** [reply]
Lua is extremely easy at the beginner levels. I know children who know Lua, mainly because of Roblox. Technically, it's a prototype language like Javascript. I'm not sure exactly what that means. I think it is how Javascript can just start with a generic object {} and you can just add member variables and functions willy nilly. I personally hate the way people program Javascript because it's so hard to read. But yeah.

Lua doesn't come with OO built-in, but it can be added using metatables and that is  very advanced programming IMO. There are OO libraries people can use. I can't remember the most popular one.

From my perspective one thing I've noticed about Lua is that the community is small and doesn't seem to last very long. Meaning, a lot of Lua libraries seem abandoned. IDK. It could be that there just are so few commercial contributors to Lua unlike something like Unity or Unreal which have big money behind them.

**Reply by magnusviri** [reply]
[Middleclass](https://github.com/kikito/middleclass) seems to be the most popular Lua class system, followed by [classic](https://github.com/rxi/classic). You can find Lua libraries [here](https://github.com/uhub/awesome-lua).

**Comment by 1010todd** [comment]
As I've stated in #9504 as long as we don't entirely replace the current data format the barrier of entry simply does not increase. And pretty much all implementation I've seen never plan to do so. This is simply expanding the ceiling of what can be done, massively. 

I think we're far too "concerned" about accessibility for plugin than we need to be. The current data format is easy enough imo, but far too limiting. Someone could very easily make a ship but as soon as they want it to do something a little special it could go straight to impossible, I've seen far too many aspiring modders disappear because what they want to do is just impossible, or rather, not possible without source code changes which is a gigantic jump in "barrier of entry" for many people. If anything an intermediate scripting language like Lua with access to more API and proper programming logic would greatly decrease the barrier to entry to vastly more features. 

So instead of being overly attached to "accessibility" we already have, why not expand the *potential* that we lack?

(Also if we're still concerned about some popular feature being only doable with scripting then consider the fact that it's even possible at all and maybe take it as hint to implement it in source code and the data.)

**Reply by Saugia** [reply]
I will note that when MZ made the issue, he did comment that it's based around content creation in general, and he didn't close it until 2 years later, but did put a label on it at the time. My guess, as with some other ambitious projects, it would have been considered as a 1.0 thing as an afterthought. If that is indeed the case and/or we feel that's appropriate (again not speaking for code because I can't), then I'd imagine the concern with accessibility would be resolved just by stating we don't want vanilla content to make use of Lua. (And, at the extreme end, what you mention in something only being able to use it, but avoiding so if possible.)

**Reply by TheGiraffe3** [reply]
... but if we say that vanilla doesn't use Lua, then people decide that it shouldn't happen because stuff that only helps plugins should not exist.
I would say (if we implement it) vanilla _can_ use Lua, to a specific extent.
Otherwise it will be taken out.

**Reply by 1010todd** [reply]
As we all know by now saying something is going to be "after 1.0" pretty much means never (or in a decade or two if people really want to split hair and take thing literally), I'd rather have it far sooner than that and have more people stick around with modding than disappearing because their ideas are straight up impossible to implement. 

And yeah, if the concern is about vanilla content using scripting then just make rules against it or something. It'd be utterly ridiculous to block off an extremely useful modding feature for something devs have absolute control over. (Also I'd argue it'd be useful for people to be able to implement cool stuffs with scripting without having to wait for coders to make it but whatever, lets argue about it after scripting become a thing.)

**Reply by 1010todd** [reply]
> ... but if we say that vanilla doesn't use Lua, then people decide that it shouldn't happen because stuff that only helps plugins should not exist. I would say (if we implement it) vanilla _can_ use Lua, to a specific extent. Otherwise it will be taken out.

Features don't need to exist *only* for vanilla to use. Why not just remove plugin support then because vanilla won't use it?

Live up to the modding support that's advertised so much. Let people make cool stuffs.

**Reply by lumbar527** [reply]
In ES, most of the time this wouldn't be used in vanilla I think as features would just get implemented in the actual game code. However, if a script could be called during a mission that would be useful for vanilla. Also, as 1010todd is saying, this game places a lot of emphasis on the fact that you can make plugins. If you can, then why is there the excuse "it won't be used for vanilla". This is the perfect thing to add. Other features that are wanted only for plugins could potentially created some using Lua.

**Reply by TheGiraffe3** [reply]
> > ... but if we say that vanilla doesn't use Lua, then people decide that it shouldn't happen because stuff that only helps plugins should not exist. I would say (if we implement it) vanilla _can_ use Lua, to a specific extent. Otherwise it will be taken out.
> 
> Features don't need to exist _only_ for vanilla to use. Why not just remove plugin support then because vanilla won't use it?
> 
> Live up to the modding support that's advertised so much. Let people make cool stuffs.

I know that, but there have been people who said otherwise.

**Reply by magnusviri** [reply]
I wouldn't pander to people who want to do stuff via modding but then disappear when they learn they can't code. It's very likely they would disappear no matter what. They'd stay around long enough to create enough of a plugin to add it to your list of plugins but then disappearing, adding a maintenance burden to the main contributors who have to keep everything working. Years ago I played Oolite, which can be modded with code. I tried some plugins and they were cool at first, but wading through them all pretty much turned me off the game because I got my hopes up but it seemed like none of the plugins worked right and a lot of them plain crashed the game. It was especially bad when the inner game mechanics changed but the plugins weren't updated. Versioning the plugins will be very important. Minetest has the same problem. It's great but it's an absolute mess. Almost all of the plugins I was interested in were abandoned and no longer worked.

I think it's important to distinguish between 2 types of plugins. There's plugins that run at update or step functions, and there's plugins that run at specific event hooks. The update/step type plugins would be required for any realtime game play, like changing how a ship behaves. The specific event hooks would be more about story line stuff. I personally think that if someone is going to change how a ship behaves, they should be doing that in c++. I know a lot of games are written in full blown Lua, so it's possible to have fast plugins. But I just think it's opening a can of worms that is best left closed. If you want to do something that affects fps do it in c++, not because of speed concerns, but because of code quality concerns. c++ requires more skill, discipline, and has a higher degree of scrutiny. If you opened this up to Lua then you lose control over what happens. I think one of the reasons this game is so good is because there is so much discipline with the existing plugin system. Adding Lua could easily destroy that and I think that should be avoided at all costs.

I still don't know the existing plugin system, so I don't know exactly what this means. I just know that unlike most games I've played, I haven't yet gotten bored of this game and it doesn't seem like it's going to happen soon. There's something really polished about this game and I'd try hard not to lose that. I guess that's why I started this discussion, because I don't know where the line is between what a Lua plugin should do either but only because I don't know the game well enough. Philosophically I believe that minimal scripting should be added. I think it's amazing you've been able to add as much plugin support as you have without coding. Keep that.

**Reply by 1010todd** [reply]
> I wouldn't pander to people who want to do stuff via modding but then disappear when they learn they can't code.

I'm not talking about people who disappear because they can't code but disappear because their plugin idea is completely impossible with current ES plugin system and their next step is now to learn C++, figure out git, figure out how to set up IDE, code C++ up to required standard, pray to god devs like the idea, rot for a year or two and hopefully not entirely lose interest long before that. I've been here long enough to see people disappear at every step of what I just said. There's way too much in the way of getting an idea into something people could experience and I hope lua could help with that.

> It was especially bad when the inner game mechanics changed but the plugins weren't updated. Versioning the plugins will be very important.

Plugin versioning is another problem of it's own, it's on the metadata side. We could make it so plugin.txt contain the game version it's supposed to work on and have the game warn you if it's outdated or unspecified (e.g. plugins prior to such update).

> Almost all of the plugins I was interested in were abandoned and no longer worked.

This is going to be a problem everywhere, even in ES some very old plugin that tried doing anything more than the most bog-standard ship and outfit won't work correctly (at some point some even crashed the game), mostly because they do some hacky thing exploiting a bug or specific behavior at the time to do something new. Also guess what, *vanilla* game also suffer from this at one point. Perhaps if there's a route available for people to more properly implement such feature than maybe it wouldn't happen as often?

Right now most normal ES plugins even older ones rarely cause crash or other major problem because ES data is mostly static, it's more of a database than actual scripting language and devs do maintain pretty good backward compatibility(probably the best of several dozen moddable games I've played). This more data-like of ES mods system also cause it to be extremely limiting especially in gameplay feature. Every little thing that requires code change that took anywhere from 1 to 5+ years since a PR for such feature exist before it's merged and most died somewhere in between before it's merged. The point is, this is way too long. Not that we can blame the devs who are all volunteers either. I think this is where plugin can supplement, if it could. I'd rather have a basically infinite amount of possibilities at cost of some stability than great stability at extreme limitations. A bunch of other games with far more complex modding system managed to have several dozens if not hundreds of mods with codes of questionable quality work together fine. I think we will manage.

**Comment by tibetiroka** [comment]
What about using normal shared libraries instead of a scripting language? That would avoid much of the development and performance overhead of integrating a scripting language, in exchange for slightly worse cross-version compatibility.

With this solution, a shared library could be compiled using the ES codebase as a dependency, negating the need to duplicate functions for script use. It would also be running native code, which has far better performance than interpreters. This is a fairly common way of modding games.

We would still need to add triggers and a provider framework to the game, but that should be easier. Each script would also have to be compiled natively, but we could create a github action setup to trivialize that (and the overall packaging of the plugin). (Though testing stuff locally would take a bit longer because of the additional compilation step.)

The big issue is ABI compatibility. This is not something that was previously considered during ES development, and could potentially break scripts quite often. Some of this can be avoided with inline namespaces, but it's likely that scripts would have to be recompiled after a couple of game releases. (Often without any actual changes to the code, so this could also be automated.)

(By the way, a big brain play we could do with this approach is to have an official plugin adding lua support, effectively bridging the two APIs. We could then even define and _automatically_ assign safety ratings to plugins, depending on the kind of scripts they support. This would preserve the performance and abilities of C++ without compromising on the security offered by lua.)

**Reply by 1010todd** [reply]
So basically have modders pre-compile the lua files and have the game load that instead of the raw script?
Sounds good to me assuming I could do it in a command or two in the shell not needing an entire IDE. I don't think it'd take that much longer to test stuffs if it's easy to compile locally.

Recompiling the scripts every other update doesn't sound too bad to me especially without any code change required, would be nice if it's more like yearly but not exactly much of a problem. Automatically assigning safety rating also sounds really nice.

**Reply by magnusviri** [reply]
Are you saying that adding code functionality would have to be compiled c++ libraries that are loaded at runtime? I don't think Lua can be precompiled. The whole point of Lua is that it isn't compiled. LuaJIT is a thing but I think it's a bit behind the main version of Lua. And I really don't think speed is the biggest problem with adding Lua support, I think the biggest problem is a polluted list of plugins on the ES webpage that don't work.

Also, all Lua scripts can be run safely without crashing by using the pcall (protected call). But a Lua script can still screw up the game state, like adding tons of credits or deleting all of the weapons on a ship.

**Reply by tibetiroka** [reply]
No, I'm saying that we could use C++ (or any compatible language) for scripting. Yes, that C++ code has to be compiled. But we could just have a C++ plugin that adds support for Lua scripts that don't need to be compiled.

**Reply by magnusviri** [reply]
This makes sense. I think it's out of scope of what I was planning on doing though. If you guys do this I'll try to incorporate what I'm doing to use what you make available.

I'm not entirely sure how this works though. I read about shared library plugins years ago and I don't remember the details. Would it be similar to what I'm doing where I find the places where I want to execute my code and write function calls to see if there's code to run?

I don't know if this is a helpful analogy and I don't know why I'm stating it but here it goes. ES is like a web browser. The built-in and plugin data files are like HTML. What I was proposing is like adding Javascript on event hooks and the DOM. But what you're proposing is more like the Adobe flash plugin.

**Comment by magnusviri** [comment]
Is there a detailed list of things people have wanted to add but the existing plugin system can't do?

My item on the list is to talk to an outside process. I started on this last night. I have an external python script that opens a file socket for reading. The Lua portion that runs in the game writes data to a socket. The python script takes the data and constructs a prompt, sends it to the LLM, and then takes the JSON result from the LLM and formats it into something ES can handle, then it saves that to a regular file. While the LLM stuff is happening ES will continue playing and will periodically check for the existence of the results file. Once it shows up it will load the results of the LLM and deal with it. I've only got the socket part working at this point. I've done lots of talking to LLM's, so I just have to repurpose my other python scripts to do that.

The hard part for me is figuring out where to put the function calls that run my Lua code, how to get all the data out of ES that I need, then how to react to the data I get back from the LLM. I am not really sure what I want it to do until I start to play it. That's actually why I'm working on the LLM part already. I am a horrible planner. I don't know what I want it to do yet.

Oh, and right now I'm trying to figure out how to support multiple Lua plugins. The code right now loads everything into one Lua state, and I don't think that's a good idea because it wasn't actually working (but maybe that was for a different reason). But if I create a state for every plugin, would that suck all the resources? I should probably ask some Lua experts.

**Reply by 1010todd** [reply]
Most of what ES plugin system is lacking is probably on the UI and gameplay side.
I haven't thought much about the UI stuffs but at least for gameplay being able to say, get and set values of ships would be a pretty big and if possible during gameplay (i.e. in flight) would be quite a massive improvement. There's going to be so many things that could be done just being able to grab some stats, do something with it on the lua side and put it back. (Also no, stop being concerned about performance until it's actually a problem.) 

There's several more things I could think of the above would be pretty decent for a starter I think.

**Reply by magnusviri** [reply]
Just thinking out loud. One idea is to have a coordinated mission. By that I mean several fleets fly to different systems and do a sweep and have them converge on a final system.  The plugin could keep track of each fleet in each system.

Also, some missions could finish on their own if you don't finish them. For example, someone could ask you to do something, and if you don't do it in time, they hire some other ship. But then they're willing to hire you for next mission in the sequence. This is mainly for missions with deadlines.

Another idea of some fun things an LLM AI could do. When asking a ship for help, it could reply with an AI generated message. I just asked a Navy Cruiser for help and it said it was too busy. It would be funny if I could try to convince them and if I failed, the Navy Cruiser could issue me a fine. Or if I asked a pirate for help they could try to take over my ship instead or open fire on me.

**Comment by magnusviri** [comment]
I've spent the last week modifying the code so that each Lua plugin has its own state. I noticed that with only one state only one plugin worked. I might've been able to get it so that each plugin worked using 1 shared state, but I thought it is probably best that they be separate from each other. This was kind of a big refactor and it took me a while to get it to work.

https://github.com/magnusviri/endless-sky/commit/480f9d0aab73cd4c7ea45190a811196c878e56b5

I've also moved most of the code out of Lua.cpp and into LuaPlugin.cpp or LuaImpl.cpp. I think the idea is that LuaPlugin deals with loading the Lua plugins and managing them, and LuaImpl exposes the ES internal data to the Lua plugins. Lua.cpp contains the hook code that the ES functions would call. Lua.cpp is pretty empty now since almost everything that was in it is now in LuaPlugin or LuaImpl.

I also changed it so that the package.path variables point to the plugin directories. This is so that plugins can include their own libraries. The paths are setup so that you can run `luarocks install --tree /path/to/endless-sky/plugins/plugin-name/scripts package-name` and it will work. Of course, that assumes the luarocks version is the same as the version that ES is built with (5.4). The plugin that my branch contains prints out the path and cpath so that you can see what they are.

https://github.com/magnusviri/endless-sky/tree/lua-0.10.8

Just to reiterate, I learned c++ a long time ago and there's way too much I don't know. I'm trying to follow the style that I've noticed from the ES files I've looked at, but I don't have a lot of experience here and I'm afraid there's things in my code that would make other people cringe. And I used AI a lot. I don't know how you guys feel about it, but I just don't have the experience to do what I'm doing on my own. I've modified the AI code a lot, but it's still pretty close to what the AI generated. And there is a lot I don't understand and so yeah. Again, I have no real hopes of getting it merged as it is. I figure if someone wants what I've made in the main branch, a lot will have to be cleaned up and I'm probably not the one who has the knowledge to do it.

**Comment by magnusviri** [comment]
Here's my first technical question in this discussion. My plugin needs to be able to save and load data. I've been looking over PlayerInfo.cpp, `PlayerInfo::Save`. Here's the conclusion I came to. Right now, `PlayerInfo::Save` calls out to several other classes to do their own saving and loading, like `Mission::Save`. So I think I need to modify `PlayerInfo::Save` to call a method named `Lua::Save`, and that will call each `LuaPlugin::Save`, and that calls a Lua function named `es_save`. If `es_save` returns text, then LuaPlugin::Save saves the text using `out.Write`.

I guess the reason I'm asking this is because I've never worked on an open source project like this. So I'm just tip toeing around. I'll go implement the code now and if I get comments I'll reply then. When I finish I'll post a link to the commit.

**Reply by magnusviri** [reply]
Nope, I'm wrong. I need to try to expose the DataWriter class to Lua and have the Lua code call the needed DataWriter functions. The biggest problem is that LuaBridge3 doesn't support overloading, so I'll have to create multiple DataWriter::Write functions, each dedicated to the data type it writes, something like `WriteCharStarToken` and `WriteStringToken` for `WriteToken(const char *a)` and `WriteToken(const std::string &a)` respectively.

**Reply by magnusviri** [reply]
I got so [incredibly close](https://github.com/magnusviri/endless-sky/commit/50e1d3520d91643086a62d988ac52e5b3af7572d). It can save, it can load, but I can't get the children. Until next time that I've got free time...

**Comment by TheGiraffe3** [comment]
Looks like Lua might help address #558.
https://github.com/endless-sky/endless-sky/issues/558#issuecomment-582558801 says other games with translations use Lua.
